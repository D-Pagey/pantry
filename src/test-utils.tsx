import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';
import React, { FC, ReactNode } from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { FirebaseContext } from './components/ProviderFirebase';

type AllTheProvidersProps = {
    children: ReactNode;
    customContext?: any;
};

const AllTheProviders: FC<AllTheProvidersProps> = ({ children, customContext }) => (
    <FirebaseContext.Provider
        value={{
            categories: [],
            deleteFoodItem: () => () => null,
            expiringCount: 0,
            isAuthed: false,
            isCheckingAuth: false,
            fridge: [],
            signOut: () => null,
            updateFridge: () => null,
            user: {
                email: null,
                name: null
            },
            ...customContext
        }}
    >
        <MemoryRouter>{children}</MemoryRouter>
    </FirebaseContext.Provider>
);

const customRender = (ui: any, customContext?: any, options?: any) =>
    render(ui, { wrapper: () => <AllTheProviders customContext={customContext}>{ui}</AllTheProviders>, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
