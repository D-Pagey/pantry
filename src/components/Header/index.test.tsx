import React from 'react';
import userEvent from '@testing-library/user-event';

import { render } from '../../test-utils';
import { UserDan } from '../../fixtures';
import { Header } from '.';

const mockHistoryBack = jest.fn();

jest.mock('react-router-dom', () => ({
    // @ts-ignore
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        goBack: mockHistoryBack
    })
}));

const context = {
    isAuthed: false,
    isCheckedAuth: false,
    user: UserDan,
};

describe('Header component', () => {
    it('should render', () => {
        const { container } = render(<Header />, context);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a title if provided', () => {
        const page = 'Add food';

        const { getByText } = render(<Header page={page} />, context);

        getByText(page);
    });

    it('should call goBack when clicked on back arrow', () => {
        const { getByTestId } = render(<Header page="Add food" />, context);

        userEvent.click(getByTestId('headerBackArrow'));

        expect(mockHistoryBack).toHaveBeenCalled();
    });

    it('should render a Notifications bell on mobile', () => {
        const authedContext = {
            ...context,
            isAuthed: true,
        };

        const { getByTestId } = render(<Header />, authedContext);
        getByTestId('header-notifications');
    });

    it('should render Notifications panel if Bell is clicked', () => {
        const authedContext = {
            ...context,
            isAuthed: true,
        };

        const { getByTestId } = render(<Header />, authedContext);
        
        userEvent.click(getByTestId('header-notifications'));

        getByTestId('notifications');
    });
});
