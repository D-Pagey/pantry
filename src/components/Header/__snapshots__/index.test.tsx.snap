// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: white;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.06);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 100px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 0 24px;
  left: 0;
  position: fixed;
  right: 0;
  top: 0;
  z-index: 2;
}

.c2 {
  color: initial;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  font-size: 25px;
  font-weight: 500;
  margin: 0;
  padding: 0 0 0 14px;
}

.c3 {
  position: relative;
  top: 2px;
  width: 32px;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c6 {
  background-color: transparent;
  border: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 0;
}

.c5 {
  position: relative;
  right: 4rem;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <a
      aria-current="page"
      class="c2 active"
      href="/"
    >
      <img
        alt="icon"
        class="c3"
        src="icon.svg"
      />
    </a>
    <a
      aria-current="page"
      class="c2 active"
      href="/"
    >
      <h4
        class="c4"
      >
        Pantry
      </h4>
    </a>
  </div>
  <div
    class="c5"
  >
    <button
      class="c6"
      type="button"
    >
      <img
        alt="notifications"
        src="bell.svg"
      />
    </button>
  </div>
</div>
`;
